### PR TITLE
[CLOUDSTACK-10323] Allow changing disk offering during volume migration 

### PR DESCRIPTION
**Pull request by rafaelweingartner**
This is a continuation of work developed on PR #2425 (CLOUDSTACK-10240), which provided root admins an override mechanism to move volumes between storage systems types (local/shared) even when the disk offering would not allow such operation. To complete the work, we will now provide a way for administrators to enter a new disk offering that can reflect the new placement of the volume. We will add an extra parameter to allow the root admin inform a new disk offering for the volume. Therefore, when the volume is being migrated, it will be possible to replace the disk offering to reflect the new placement of the volume.

The API method will have the following parameters:

* storageid (required)
* volumeid (required)
* livemigrate(optional)
* newdiskofferingid (optional) – this is the new parameter

The expected behavior is the following:

* If “newdiskofferingid” is not provided the current behavior is maintained. Override mechanism will also keep working as we have seen so far.
* If the “newdiskofferingid” is provided by the admin, we will execute the following checks
    * new disk offering mode (local/shared) must match the target storage mode. If it does not match, an exception will be thrown and the operator will receive a message indicating the problem.
    * we will check if the new disk offering tags match the target storage tags. If it does not match, an exception will be thrown and the operator will receive a message indicating the problem.
    * check if the target storage has the capacity for the new volume. If it does not have enough space, then an exception is thrown and the operator will receive a message indicating the problem.
    * check if the size of the volume is the same as the size of the new disk offering. If it is not the same, we will ALLOW the change of the service offering, and a warning message will be logged.

We execute the change of the Disk offering as soon as the migration of the volume finishes. Therefore, if an error happens during the migration and the volume remains in the original storage system, the disk offering will keep reflecting this situation.

Attached is a screenshot to show how the migrate option will look like:
changeintheui

### CODE REVIEW

**Comment by borisstoyanov** [comment]
thanks for this PR @rafaelweingartner, are you looking into adding a marvin tests for this ?
I'll wait on your response and kick the smoketests.

**Comment by rafaelweingartner** [comment]
I added unit test for the code I changed. I was not planning to create a Marvin tests though.
If you guys think it is a requirement, I can look into it.

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
ok @rafaelweingartner, let me run some tests

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@rafaelweingartner I've created a marvin test but I'm not able to push to your repo. How can I contribute?

**Comment by rafaelweingartner** [comment]
Thanks!
You can open a PR against my branch, then I can merge it here.

**Comment by borisstoyanov** [comment]
I'm not sure I'm doing this right.. but git says there's nothing to compare.. 
screen shot 2018-03-15 at 12 51 51
any idea? @rafaelweingartner

**Comment by rafaelweingartner** [comment]
Did you pushed your changes to your branch in the remote repository?
I am checking your branch, but there is no new commit there.

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
debian packaging failed and so did jenkins and travis!

**Comment by rafaelweingartner** [comment]
Yes. However, that is a consequence of the last merge and is fixed with: #2491

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by rafaelweingartner** [comment]
@DaanHoogland now everything is ok again.

**Comment by DaanHoogland** [comment]
thanks @rafaelweingartner , didn't notice the relation

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 1]
```diff
@@ -2097,26 +2079,102 @@ public Volume migrateVolume(MigrateVolumeCmd cmd) {
...
 
-    private Volume orchestrateMigrateVolume(long volumeId, long destPoolId, boolean liveMigrateVolume) {
-        VolumeVO vol = _volsDao.findById(volumeId);
-        assert (vol != null);
-        StoragePool destPool = (StoragePool)dataStoreMgr.getDataStore(destPoolId, DataStoreRole.Primary);
-        assert (destPool != null);
+    /**
+     * Retrieve the new disk offering UUID that might be sent to replace the current one in the volume being migrated.
+     * If no disk offering UUID is provided we return null. Otherwise, we perform the following checks.
+     * <ul>
```
Comment (nitin-maharana) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: @rafaelweingartner, I think you have added few HTML tags by mistake here.

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 2]
```diff
@@ -2097,26 +2079,102 @@ public Volume migrateVolume(MigrateVolumeCmd cmd) {
...
+        }
+        Account caller = CallContext.current().getCallingAccount();
+        _accountMgr.checkAccess(caller, newDiskOffering);
+        return newDiskOffering;
+    }
+
+    /**
+     * Performs the validations required for replacing the disk offering while migrating the volume of storage. If no new disk offering is provided, we do not execute any validation.
+     * If a disk offering is informed, we then proceed with the following checks.
+     * <ul>
```
Comment (nitin-maharana) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: Same here.

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java (hunk 1 above)
Comment (rafaelweingartner) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: It is not by mistake. It is on purpose to format the java documentation. Using them you are able to create javadocs as this one:
javadoc

Comment (nitin-maharana) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: Got it. Thanks.

**Comment by rafaelweingartner** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by rafaelweingartner** [comment]
@borisstoyanov are these errors something that I should take a look into? I noticed that the test you included is failing as well.

I look at the error of the new test, and it seems you are trying to change the service offering to a new one that does not have the tags of the target storage.

`Target Storage [id=0818cecf-5f99-3b1e-beb2-5f0669c7b5ba] tags [storage.overprovisioning.factor] does not match new disk offering [id=a2992784-662b-4cad-b833-38c46136f97e] tags [null].'}, accountid : u'9f16aecc-2c1d-11e8-8460-06d348010705'}\n"]`

**Comment by borisstoyanov** [comment]
@rafaelweingartner let me run this again, don't know why it appeared.
@blueorangutan test

**Comment by rafaelweingartner** [comment]
Now the errors changed, it increased the number of them. However, the new test case that was added keeps failing.

**Comment by borisstoyanov** [comment]
@rafaelweingartner if we add this in the marvin tests it should clean before migration, can you just add it before it tries to migrate?
```
        if hasattr(pool, 'tags'):
            StoragePool.update(self.apiclient, id=pool.id, tags="")
```

**Comment by rafaelweingartner** [comment]
@borisstoyanov I ammended the code as you suggested.

**Comment by rafaelweingartner** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rafaelweingartner** [comment]
@borisstoyanov with that change in the test case, now the test we created for this PR passed. Some tests failed, but they are different from the last run.

**Comment by borisstoyanov** [comment]
yup, let me run it again @blueorangutan test

**Comment by PaulAngus** [comment]
@blueorangutan test

**Comment by rafaelweingartner** [comment]
@nitin-maharana are you LGTM here?

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 3]
```diff
@@ -2097,26 +2079,102 @@ public Volume migrateVolume(MigrateVolumeCmd cmd) {
...
+            DiskOfferingVO oldDiskOffering = this._diskOfferingDao.findById(volume.getDiskOfferingId());
+            s_logger.warn(String.format(
+                    "You are migrating a volume [id=%s] and changing the disk offering[from id=%s to id=%s] to reflect this migration. However, the sizes of the volume and the new disk offering are different.",
+                    volume.getUuid(), oldDiskOffering.getUuid(), newDiskOffering.getUuid()));
+        }
 
+    }
+
+    /**
+     *  Retrieve the storage pool tags as a {@link String}. If the storage pool does not have tags we return a null value.
```
Comment (GabrielBrascher) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: I think that  "Retrieves" would best fit here.

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 4]
```diff
@@ -2097,26 +2079,102 @@ public Volume migrateVolume(MigrateVolumeCmd cmd) {
...
+        return orchestrateMigrateVolume(vol, destPool, liveMigrateVolume, newDiskOffering);
     }
 
-    private Volume orchestrateMigrateVolume(long volumeId, long destPoolId, boolean liveMigrateVolume) {
-        VolumeVO vol = _volsDao.findById(volumeId);
-        assert (vol != null);
-        StoragePool destPool = (StoragePool)dataStoreMgr.getDataStore(destPoolId, DataStoreRole.Primary);
-        assert (destPool != null);
+    /**
+     * Retrieve the new disk offering UUID that might be sent to replace the current one in the volume being migrated.
```
[the remark this answers is not in the document]
Comment (rafaelweingartner) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java]: If you read as "it (the method) retrieves the new .....". Then, yes, I guess it should be "Retrieves".
I will change that.

**Pull request merged**